websockets/README.md

```diff
@@ -0,0 +1,166 @@
+
+### Policy support 
+With a WebSockets connection, you can only use the [Verify API Key](https://cloud.google.com/apigee/docs/api-platform/reference/policies/verify-api-key-policy) and [OAuthV2](https://cloud.google.com/apigee/docs/api-platform/reference/policies/oauthv2-policy) policies in your API proxy. All other policies are ignored.
+
+### Revoking the connection 
+
+The WebSockets connection is closed when:
+
+* The proxy endpoint receives a request without an API key or OAuth token.
+* The proxy endpoint receives a request with an expired or invalid API key or OAuth token.
+* The WebSockets connection times out.
+
+## Create a websockets backend in Cloud Run
+
+Create a Docker image:
+In Cloud Shell, clone the websockets repo 
+
+```sh
+export PROJECT=$GOOGLE_CLOUD_PROJECT
+mkdir websockets-backend
+cd websockets-backend
+git clone https://github.com/websockets/websocket-echo-server.git
+cd websocket-echo-server
+```
+
+Build the Docker image and then list the images created in CloudShell:
+
+```sh
+docker build .
+docker images -a
+```
+
+Save the imageid from above to an environment variable and run the Docker image locally:
+
+```sh
+export DOCKER_IMAGE=<imageid>
+docker run -e BIND_PORT=8080 --expose 8080 -d -p 8080:8080 $DOCKER_IMAGE
+```
+
+Open a second tab in your CloudShell and connect to your localhost to verify that the server is working. (We will be installing wscat for this test):
+
+```sh
+npm install -g wscat
+wscat --connect ws://localhost:8080
+```
+
+You should be able to type anything and receive an echo response from the server:
+
+```
+Connected (press CTRL+C to quit)
+> hello
+< hello
+```
+
+Once you verified that the image is working, you can close the second tab.
+Now let's create a global Artifact Registry and configure Docker to use the Artifact Registry:
+
+```sh
+gcloud artifacts repositories create websocketsdemo \
+    --repository-format=Docker \
+    --location=us \
+    --description="Websockets Demo repo"
+
+gcloud auth configure-docker us-docker.pkg.dev
+```
+
+Tag the Docker image, Push the Docker image, and create a Cloud Run service from the Docker image:
+Optional: Change REGION= to a region closest to your Apigee instance
+
+```sh
+export REGION=us-central1
+docker tag $DOCKER_IMAGE us-docker.pkg.dev/$PROJECT/websocketsdemo/websockets-echo-server
+docker push us-docker.pkg.dev/$PROJECT/websocketsdemo/websockets-echo-server
+gcloud run deploy \
+--set-env-vars BIND_PORT=8080 \
+websockets-echo-server \
+--image us-docker.pkg.dev/$PROJECT/websocketsdemo/websockets-echo-server \
+ --platform managed --allow-unauthenticated \
+--region $REGION
+```
+
+If this command is successful it should deploy the websockets server to Cloud Run and return the service URL.
+
+Go ahead and save the Service URL, since it will be used in the next step when we configure the Apigee proxy.
+
+```
+export CLOUD_RUN_SERVICE_URL=<Cloud Run service endpoint from above without the "https://" prefix>
+```
+
+## Prerequisites
```

Comment:
Mention they need to enable Artifact Registry and Cloud Run, plus the associated permissions